### PR TITLE
Use lastSyncedGlobalCheckpoint in deletion policy

### PR DESCRIPTION
**Pull request by dnhatn**
Today we use the in-memory global checkpoint from SequenceNumbersService
to clean up unneeded commit points, however the latest global checkpoint
may haven't fsynced to the disk yet. If the translog checkpoint fsync
failed and we already use a higher global checkpoint to clean up commit
points, then we may have removed a safe commit which we try to keep for
recovery.

This commit updates the deletion policy using lastSyncedGlobalCheckpoint
from Translog rather the in memory global checkpoint.

Relates #27606

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 1]
```diff
@@ -385,22 +385,6 @@ protected void readBytes(ByteBuffer targetBuffer, long position) throws IOExcept
         Channels.readFromFileChannelWithEofException(channel, position, targetBuffer);
     }
 
-    private static Checkpoint writeCheckpoint(
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: why is this relevant? I rather not touch files that aren't relevant for the PR (feel free to push such a change as removal of dead code without a PR)

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 2]
```diff
@@ -4195,34 +4194,44 @@ public long getGlobalCheckpoint() {
...
-            int numDocs = scaledRandomIntBetween(10, 100);
-            for (int docId = 0; docId < numDocs; docId++) {
-                ParseContext.Document document = testDocumentWithTextField();
-                document.add(new Field(SourceFieldMapper.NAME, BytesReference.toBytes(B_1), SourceFieldMapper.Defaults.FIELD_TYPE));
-                engine.index(indexForDoc(testParsedDocument(Integer.toString(docId), null, document, B_1, null)));
-                if (frequently()) {
-                    globalCheckpoint.set(randomIntBetween(
-                        Math.toIntExact(engine.seqNoService().getGlobalCheckpoint()),
-                        Math.toIntExact(engine.seqNoService().getLocalCheckpoint())));
+        engine = new InternalEngine(config(indexSettings, store, createTempDir(), NoMergePolicy.INSTANCE, null), seqNoServiceSupplier) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: why did we lose the try with resources clause?

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 3]
```diff
@@ -4195,34 +4194,44 @@ public long getGlobalCheckpoint() {
...
+            document.add(new Field(SourceFieldMapper.NAME, BytesReference.toBytes(B_1), SourceFieldMapper.Defaults.FIELD_TYPE));
+            engine.index(indexForDoc(testParsedDocument(Integer.toString(docId), null, document, B_1, null)));
+            if (frequently()) {
+                globalCheckpoint.set(randomIntBetween(
+                    Math.toIntExact(engine.seqNoService().getGlobalCheckpoint()),
+                    Math.toIntExact(engine.seqNoService().getLocalCheckpoint())));
+                engine.getTranslog().sync();
+            }
+            if (frequently()) {
+                final long lastSyncedGlobalCheckpoint = engine.getTranslog().getLastSyncedGlobalCheckpoint();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: strictly speaking I think we need to read this from disk after the flush - i.e., make sure that what's on disk is OK.

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 4]
```diff
@@ -4195,34 +4194,44 @@ public long getGlobalCheckpoint() {
...
-                engine.index(indexForDoc(testParsedDocument(Integer.toString(docId), null, document, B_1, null)));
-                if (frequently()) {
-                    globalCheckpoint.set(randomIntBetween(
-                        Math.toIntExact(engine.seqNoService().getGlobalCheckpoint()),
-                        Math.toIntExact(engine.seqNoService().getLocalCheckpoint())));
+        engine = new InternalEngine(config(indexSettings, store, createTempDir(), NoMergePolicy.INSTANCE, null), seqNoServiceSupplier) {
+            @Override
+            protected void commitIndexWriter(IndexWriter writer, Translog translog, String syncId) throws IOException {
+                // The global checkpoint is advanced but not fsynced yet.
+                final long lagging = seqNoService().getLocalCheckpoint() - seqNoService().getGlobalCheckpoint();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: I don't follow this - why do we need to query the global checkpoint from the seqNoService? we alread have access to globalCheckpoint ?

I think what you mean here is something like this - no if no rarely:

```
 // Advance the global checkpoint during the flush to create a lag between what's persisted in the translog (and is visible for CombinedDeletionPolicy) and what's in memory in the SequenceServices
globalCheckpoint.set(randomLongBetween(globalCheckpoint.get(), seqNoService().getLocalCheckpoint()));
```

wdyt?

**Review by bleskes — CHANGES_REQUESTED** [review]
Thx Nhat. I left some comments

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 2 above)
Comment (dnhatn) [reply core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: We close the engine in tearDown but I put the try-clause back.

**Comment by dnhatn** [comment]
@bleskes I have addressed your comments. Would you please take a look? Thank you.

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 2 above)
Comment (bleskes) [reply core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: I see - we reuse the same engine variable. I think it's cleaner to have it contained.

**Review by bleskes — APPROVED** [review]
LGTM. Thx @dnhatn

**Comment by dnhatn** [comment]
Thanks @bleskes.

**Comment by dnhatn** [comment]
Backported in https://github.com/elastic/elasticsearch/pull/27866

**Pull request merged**